XMLRPC: Allow to fetch custom fields for testplan testcases

janLo opened the pull request:
We use custom fields to match testlink entries to automated testcases.
At the moment the reporting tool has to place one request per testcase
in a testplan to get the custom field value, which is very slow.

With this we allow the user to request custom fields when he asks for
all testcases in a testplan. the parameter is "customfields" and it can
be either "true" for all custom fields or an array with the requested
custom field names.

fmancardi commented on lib/api/xmlrpc/v1/xmlrpc.class.php:
```diff
@@ -2687,9 +2690,73 @@ public function getTestCasesForTestPlan($args) {
+            null,null, $project_id);
+        $filtered_cf = array();
+
+        foreach ($cf as $cf_id => $cfield) {
```
please extract the invariant from the loop  => is_array($cf_options), do the check using the function call to is_array() just once outside the loop saving result in a variable

fmancardi commented:
Please format code following the testlink standard => indent to 2 not 4, and do not leave space between operator ! like in **! in_array($cfield['name'], $cf_options))**

janLo commented:
I updated the PR regarding your requests.

Regarding the "testlink standard" - i this anywhere specified (or even better: is there a vonfiguration for a fomatting tool somewhere)? It seems really cumbersome to manually format code towards a specific standard without having the standard or a tool implementing it.

fmancardi commented:
The standard is not completed documented, and unfortunately someone in the
past has changed it with a PR for this piece of code.
- indentation 2 with spaces
- no space between operator and expression

In your case we are talking about 40 lines of code then IMHO can not as
cumbersome as seems


On Mon, Jul 1, 2019 at 11:03 AM Jan Losinski <notifications@github.com>
wrote:

> I updated the PR regarding your requests.
>
> Regarding the "testlink standard" - i this anywhere specified (or even
> better: is there a vonfiguration for a fomatting tool somewhere)? It seems
> really cumbersome to manually format code towards a specific standard
> without having the standard or a tool implementing it.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/TestLinkOpenSourceTRMS/testlink-code/pull/203?email_source=notifications&email_token=AABZRMKCP2SZIMRTOFVBUG3P5HB6TA5CNFSM4H4FIVG2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODY5PNTY#issuecomment-507180751>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABZRMIOAO5KYXU3GMALLE3P5HB6TANCNFSM4H4FIVGQ>
> .
>


-- 
Francisco Mancardi

janLo commented:
You have to take into account, that I (or whoever the contibutor might be) very rarely use PHP and never touches the Testlink codebase exept for this one occasion where a specific feature is needed.

I misunderstook you with the two spaces - and used it only for the '=>' lines - shall I again reformat it contrary to the rest of the file?

I would suggest to use a common formatter instead of a special format and let it run as an automatic check on every PR automatically (for example as a Travis job) - then you could save a lot of time in review-roundtrips by having a fail-fast check and a tool for the contributor to fix the issues.

fmancardi commented:
ok, thanks

On Mon, Jul 1, 2019 at 11:14 AM Jan Losinski <notifications@github.com>
wrote:

> You have to take into account, that I (or whoever the contibutor might be)
> very rarely use PHP and never touches the Testlink codebase exept for this
> one occasion where a specific feature is needed.
>
> I misunderstook you with the two spaces - and used it only for the '=>'
> lines - shall I again reformat it contrary to the rest of the file?
>
> I would suggest to use a common formatter instead of a special format and
> let it run as an automatic check on every PR automatically (for example as
> a Travis job) - then you could save a lot of time in review-roundtrips by
> having a fail-fast check and a tool for the contributor to fix the issues.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/TestLinkOpenSourceTRMS/testlink-code/pull/203?email_source=notifications&email_token=AABZRMPK37LM2EE4I7NCUODP5HDIZA5CNFSM4H4FIVG2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODY5QNJA#issuecomment-507184804>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABZRMKLYCOM3XYPI5AS4O3P5HDIZANCNFSM4H4FIVGQ>
> .
>


-- 
Francisco Mancardi

janLo commented:
What are the next steps now to get this feature merged?

fmancardi commented:
if you have done the minimal changes I've requested I'm going to integrate
in next days

On Tue, Jul 2, 2019 at 10:41 AM Jan Losinski <notifications@github.com>
wrote:

> What are the next steps now to get this feature merged?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/TestLinkOpenSourceTRMS/testlink-code/pull/203?email_source=notifications&email_token=AABZRMIXJWT5TTZGHWQJVC3P5MIFHA5CNFSM4H4FIVG2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZARC5I#issuecomment-507580789>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABZRMOLGM24BYDY6H7YOGDP5MIFHANCNFSM4H4FIVGQ>
> .
>


-- 
Francisco Mancardi

janLo commented:
I've refactored the invariant and changed the spaces between the unary operators. I'm still a bit confused about the indentation as the whole file uses 4 spaces - I don't really see which bits should be indented with just two spaces.

janLo commented:
Thanks!
